Bluetooth: controller: nRF radio driver use single timer

ioannisg opened the pull request:
The aim of this work is to free a TIMER instance used by the Radio driver in the BLE controller.
The user will be able to select at compile-time, whether she would like to use this optimization, or not.

@cvinayak FYI

ioannisg commented on subsys/bluetooth/controller/hal/nrf5/radio/radio_nrf5_ppi.h:
```diff
@@ -125,6 +125,44 @@ static inline void hal_radio_enable_on_tick_ppi_config_and_enable(u8_t trx)
+ * wire the RADIO EVENTS_END event to the
+ * EVENT_TIMER TASKS_CLEAR task.
+ *
+ * Note: in nRF52X PPI 5 is forked for both capturing and clearing timer
```
Effectively, the macros used for the FORKED task are the same as for the PPI for capturing the END time, so we will experience some minor performance degradation, since we will be trying to set the EEP register twice, with the same value.
This is the small price to pay, to get consistency with the rest of PPI configuration, and ease-of-reading.

SebastianBoe commented:
You can put DNM or WIP in the title to communicate that is not in a merge-able state. :)

cvinayak commented on subsys/bluetooth/controller/hal/nrf5/radio/radio.c:
```diff
@@ -24,8 +24,14 @@
+
 void isr_radio(void)
 {
+	/* On packet END event, increment last packet end time value. */
```
This function is used for Radio ISR sharing amongst BLE states/roles, any role specific code should be in the functions set using radio_isr_set()

ioannisg replied:
@cvinayak I am not sure I understood your comment here. Could you elaborate a bit more?

If I understood correctly: you would like the code, here, to be moved to the ISR callbacks in the roles. However, this would mean that this code would be replicated several times. I wonder if this optimal.

cvinayak replied:
Move this inside radio_is_done() interface, and conditional compile (both memory used and code). You will avoid redundant branching instructions.

cvinayak commented on subsys/bluetooth/controller/hal/nrf5/radio/radio.c:
```diff
@@ -791,22 +800,36 @@ void radio_tmr_end_capture(void)
 		HAL_RADIO_END_TIME_CAPTURE_PPI_TASK;
-	NRF_PPI->CHENSET = HAL_RADIO_END_TIME_CAPTURE_PPI_ENABLE;
+
+	HAL_EVENT_TIMER_CLEAR_PPI_REGISTER_EVT = HAL_EVENT_TIMER_CLEAR_PPI_EVT;
```
My origin code before refactor used SOC registers directly for the reason I dont have to remember or read some other header file to know what is the CLEAR_PPI_EVT ?... ok... after some greping its the NRF_RADIO->EVENTS_END;

So now you are clearing a free running timer, but keeping track of the microseconds in a local variable. Clearing the timer would add a +/- 1 us drift?

ioannisg replied:
Clearing the timer, actually, removes jitter/drift for HCTO, PA/LNA and TIFS calculations.
It, however, introduces a drift of the conception of the absolute time inside the event. I wonder whether the absolute time is used anywhere.

ioannisg commented on subsys/bluetooth/controller/ll_sw/ctrl.c:
```diff
@@ -7998,7 +7998,8 @@ static void event_slave(u32_t ticks_at_expire, u32_t remainder, u16_t lazy,
 #if defined(CONFIG_BT_CTLR_PROFILE_ISR) || \
-    defined(CONFIG_BT_CTLR_GPIO_PA_PIN)
+    defined(CONFIG_BT_CTLR_GPIO_PA_PIN) || \
+    defined(CONFIG_BT_CTLR_SW_SWITCH_SINGLE_TIMER)
```
this was necessary, since peripheral role does not capture the END event timestamp on the first packet, otherwise.

cvinayak commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -358,6 +358,14 @@ config BT_CTLR_TIFS_HW
+	depends on (!BT_CTRL_TIFS_HW) && SOC_SERIES_NRF52X
+	default y
+	help
+		Implement the tIFS Trx SW switch with the same TIMER instance,
```
tab... then 2 spaces under the help.

cvinayak commented on subsys/bluetooth/controller/hal/nrf5/radio/radio.c:
```diff
@@ -603,12 +609,13 @@ void radio_tmr_status_reset(void)
 			HAL_SW_SWITCH_TIMER_S8_DISABLE_PPI_DISABLE |
-#endif /* !CONFIG_BT_CTLR_TIFS_HW */
-#endif /* CONFIG_SOC_NRF52840 */
+#endif /* !defined(CONFIG_BT_CTLR_TIFS_HW) */
```
remove these changes to comments.

cvinayak commented on subsys/bluetooth/controller/hal/nrf5/radio/radio.c:
```diff
@@ -24,8 +24,14 @@
 
 static radio_isr_fp sfp_radio_isr;
 
+static u32_t last_pkt_end_time_us;
```
rename to "last_pdu_end_us", pkt term be used for L2CAP

cvinayak commented on subsys/bluetooth/controller/hal/nrf5/radio/radio.c:
```diff
@@ -645,7 +652,10 @@ u32_t radio_tmr_start(u8_t trx, u32_t ticks_start, u32_t remainder)
 
 	hal_radio_enable_on_tick_ppi_config_and_enable(trx);
 
+	last_pkt_end_time_us = 0;
```
I shall see all new code wrapped around the conditional compilation.

ioannisg commented on subsys/bluetooth/controller/hal/nrf5/radio/radio_nrf52832.h:
```diff
@@ -9,7 +9,7 @@
 #define SW_SWITCH_TIMER_EVTS_COMP_BASE 0
 #define SW_SWITCH_TIMER_TASK_GROUP_BASE 0
-#endif
+#endif /* !defined(CONFIG_BT_CTLR_TIFS_HW) */
```
minor style improvement

ioannisg commented on subsys/bluetooth/controller/hal/nrf5/radio/radio_nrf52840.h:
```diff
@@ -10,7 +10,7 @@
 #define SW_SWITCH_TIMER_EVTS_COMP_S2_BASE 2
 #define SW_SWITCH_TIMER_TASK_GROUP_BASE 0
-#endif
+#endif /* !defined(CONFIG_BT_CTLR_TIFS_HW) */
```
minor style improvement

carlescufi commented:
@cvinayak can you take another look at this one?

carlescufi commented on subsys/bluetooth/controller/hal/nrf5/radio/radio.c:
```diff
@@ -24,6 +24,14 @@
 
+#if defined(CONFIG_BT_CTLR_SW_SWITCH_SINGLE_TIMER)
+static u32_t last_pdu_end_us;
+static inline void increment_last_pkt_end_time(void)
```
this seems a bit long and doesn't match the variable name, how about `inc_last_pdu_end()`?

ioannisg replied:
OK!

carlescufi commented on subsys/bluetooth/controller/hal/nrf5/radio/radio.c:
```diff
@@ -792,21 +818,41 @@ void radio_tmr_end_capture(void)
+#endif /* defined(CONFIG_BT_CTLR_SW_SWITCH_SINGLE_TIMER) */
 }
 
+static u32_t tmr_sample_val;
```
we tend to define globals at the top of the file

ioannisg replied:
agree with the general concept, of course. However:
1. this variable is only used in these two functions.
2. if we move to top, we need _yet_ another #ifdef CONFIG_BT_CTLR_SW_SWITCH_SINGLE_TIMER
3. there is already another static variable, sw_tifs_toggle that is defined right before it is used.
So, @carlescufi I would recommend to reduce noise, and keep it here.

carlescufi replied:
Ok, makes sense.

ioannisg commented:
rebased

cvinayak commented on subsys/bluetooth/controller/hal/nrf5/radio/radio.c:
```diff
@@ -890,27 +916,52 @@ void radio_tmr_end_capture(void)
 void radio_tmr_sample(void)
 {
-	EVENT_TIMER->TASKS_CAPTURE[3] = 1;
+	u32_t cc;
```
Missing conditional compilation.

Keep the old implementation as is when the new feature is not used, it has lot lesser influence to measurement.

cvinayak commented on subsys/bluetooth/controller/hal/nrf5/radio/radio.c:
```diff
@@ -890,27 +916,52 @@ void radio_tmr_end_capture(void)
 u32_t radio_tmr_sample_get(void)
 {
-	return EVENT_TIMER->CC[3];
+	return tmr_sample_val;
```
Missing conditional compilation for old implementation.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ctrl.c:
```diff
@@ -608,13 +608,22 @@ static inline u32_t empty_pkt_us_get(u8_t phy)
 
+static inline u32_t tifs_tmr_base_get(void)
+{
+#if defined(CONFIG_BT_CTLR_SW_SWITCH_SINGLE_TIMER)
```
This is a HAL option, and should not be used in ctrl.c.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ctrl.c:
```diff
@@ -8000,7 +8009,8 @@ static void event_slave(u32_t ticks_at_expire, u32_t remainder, u16_t lazy,
 #if defined(CONFIG_BT_CTLR_PROFILE_ISR) || \
-    defined(CONFIG_BT_CTLR_GPIO_PA_PIN)
+    defined(CONFIG_BT_CTLR_GPIO_PA_PIN) || \
+	defined(CONFIG_BT_CTLR_SW_SWITCH_SINGLE_TIMER)
```
This is nRF HAL specific and should not be in ctrl.c

ioannisg committed:
Bluetooth: controller: symbolic name for CC used in ISR profiling

This commit adds a symbolic name for the event timer
CC offset that is used to sample the event timer during
ISR profiling.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>
Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

ioannisg committed:
Bluetooth: controller: conditionally reuse CC for ISR profiling

This commit adds the option of conditionally reuse an existing
CC register of the event timer for timer sampling used in ISR
profiling, in case SW tIFS switching is implemented based on the
event timer.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>
Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

ioannisg committed:
Bluetooth: controller: sw-switch with the event timer (prepare commit)

This commit implements minor refactoring of radio.c and ctrl.c,
to prepare for adding the implementation of sw-switch based on
the event timer.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
Bluetooth: controller: implement sw-switch with the event timer

This commit contributes the implementation of the sw-switch for
tIFS in nRF5 Radio using the EVENT_TIMER instead of a dedicated
TIMER instance. A Kconfig configuration is added so the user can
select whether to use the EVENT_TIMER for the tIFS switch or use
a dedicated TIMER instance.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>
Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

ioannisg commented:
- Split into 4 commits
- Separating refactor (prepare) commits from commits with added implementation (conditional)
@cvinayak FYI

cvinayak committed:
Bluetooth: controller: refactor radio_nrf5* headers

Refactor radio_nrf5* header files to group together
definitions and have less #if-#else-#endif.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
@ioannisg, I have some pending refactoring that conflicts these commits, I am half-way resolving them, will include those commits in this PR before I push them here this evening.

ioannisg committed:
Bluetooth: controller: inline implementation of radio-enable-on-tIFS

This commit implements a minor refactoring of the implementation
for enabling the Radio on TIFS expiration. It inlines a function
that is defined in the local radio_nrf5_ppi header file.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

cvinayak commented:
@ioannisg and @carlescufi updated the PR with my changes.

ioannisg commented:
@cvinayak I went through your refactoring; looks ok.

I also included a (very minor) refactoring commit on top of the PR. Pls take a quick look.
